WalletWasabi/Transactions/AllTransactionStore.cs

```diff
@@ -0,0 +1,239 @@
+				Network = Guard.NotNull(nameof(network), network);
+
+				MempoolStore = new TransactionStore();
+				ConfirmedStore = new TransactionStore();
+				Lock = new object();
+
+				var mempoolWorkFolder = Path.Combine(WorkFolderPath, "Mempool");
+				var confirmedWorkFolder = Path.Combine(WorkFolderPath, "ConfirmedTransactions");
+
+				var initTasks = new[]
+				{
+					MempoolStore.InitializeAsync(mempoolWorkFolder, Network, $"{nameof(MempoolStore)}.{nameof(MempoolStore.InitializeAsync)}"),
+					ConfirmedStore.InitializeAsync(confirmedWorkFolder, Network, $"{nameof(ConfirmedStore)}.{nameof(ConfirmedStore.InitializeAsync)}")
+				};
+
+				await Task.WhenAll(initTasks).ConfigureAwait(false);
+				EnsureConsistency();
+
+				if (ensureBackwardsCompatibility)
+				{
+					EnsureBackwardsCompatibility();
+				}
+			}
+		}
+
+		private void EnsureBackwardsCompatibility()
+		{
+			try
+			{
+				// Before Wasabi 1.1.7
+				var networkIndependentTransactionsFolderPath = Path.Combine(EnvironmentHelpers.GetDataDir(Path.Combine("WalletWasabi", "Client")), "Transactions");
+				if (Directory.Exists(networkIndependentTransactionsFolderPath))
+				{
+					var oldTransactionsFolderPath = Path.Combine(networkIndependentTransactionsFolderPath, Network.Name);
+					if (Directory.Exists(oldTransactionsFolderPath))
+					{
+						lock (Lock)
+						{
+							foreach (var filePath in Directory.EnumerateFiles(oldTransactionsFolderPath))
+							{
+								try
+								{
+									string jsonString = File.ReadAllText(filePath, Encoding.UTF8);
+									var allWalletTransactions = JsonConvert.DeserializeObject<IEnumerable<SmartTransaction>>(jsonString)?.OrderByBlockchain() ?? Enumerable.Empty<SmartTransaction>();
+									foreach (var tx in allWalletTransactions)
+									{
+										AddOrUpdateNoLock(tx);
+									}
+
+									// ToDo: Uncomment when PR is finished.
+									// File.Delete(filePath);
+								}
+								catch (Exception ex)
+								{
+									Logger.LogTrace(ex);
+								}
+							}
+
+							// ToDo: Uncomment when PR is finished.
+							// Directory.Delete(oldTransactionsFolderPath, recursive: true);
+						}
+					}
+
+					// ToDo: Uncomment when PR is finished.
+					//// If all networks successfully migrated, too, then delete the transactions folder, too.
+					//if (!Directory.EnumerateFileSystemEntries(networkIndependentTransactionsFolderPath).Any())
+					//{
+					//	Directory.Delete(networkIndependentTransactionsFolderPath, recursive: true);
+					//}
+				}
+			}
+			catch (Exception ex)
+			{
+				Logger.LogWarning("Backwards compatibility could not be ensured.");
+				Logger.LogWarning(ex);
+			}
+		}
+
+		#endregion Initializers
+
+		#region Modifiers
+
+		private void AddOrUpdateNoLock(SmartTransaction tx)
+		{
+			var hash = tx.GetHash();
+
+			if (tx.Confirmed)
+			{
+				if (MempoolStore.TryRemove(hash, out SmartTransaction found))
+				{
+					found.TryUpdate(tx);
+					ConfirmedStore.TryAddOrUpdate(found);
```

Comment:
Maybe I am commenting on the wrong line, but - if it is the same as it is stored, the update operation is not required.

Reply:
Same reply as here: https://github.com/zkSNACKs/WalletWasabi/pull/2407#discussion_r332639190

Two transactions are equal if they have the same hash. There may be differences between what labels they have, when they were first seen, etc... Such inconsistencies are what update is resolving.

Reply:
Ack.